OmniauthCallbacksController Changes: Handle Additional User Scenarios and Improve Underlyling Tests

aaronskiba opened the pull request:
Fixes # Not a fix, but this PR helps address https://github.com/portagenetwork/roadmap/issues/1067

# Changes proposed in this PR:
## Handle Additional User Scenarios
- Add flash message for users signing in via CILogon
  - If the user's name is 'generic' (i.e. it matches the fallback values used for accounts created via SSO), then a flash warning is rendered, informing the user to update their profile. Otherwise, a 'Signed in successfully.' flash notice is rendered.
- Handle attempts to re-link already linked external credentials
  - This scenario is not possible through normal UI interaction, but a user could trigger it by manipulating the frontend.

## Improve Underlyling Tests
- Perform a complete rework of the corresponding OmniauthCallbacksController test file
  - Add additional test scenarios, including those corresponding to the aforementioned additional user scenarios.

aaronskiba committed:
Add SSO sign-in flash warning for generic name

This change adds a flash message when a user signs in via CILogon.
- If the user's name is 'generic' (i.e. it matches the fallback values used for accounts created via SSO), then a flash warning is rendered, informing the user to update their profile. Otherwise, a 'Signed in successfully.' flash notice is rendered.

momo3404 commented:
Manual testing steps:
1. Sign in with SSO then go to Edit Profile and change first and last name to "First name" and "Last name"
2. Logout and then try and signing in with SSO again. Should see this:
![image](https://github.com/user-attachments/assets/d7b1ba40-f069-47e2-864a-9473f5ea6d8f)

aaronskiba committed:
Fix test to align with commit # 10ebe397b

After the changes within 10ebe397bbf965f58cd766d2d38d8b153de5b64f, we now expect a `flash[:notice]` with the message 'Signed in successfully.' to render.

aaronskiba committed:
Redo OmniauthCallbacksController tests

These changes improve/fix the tests within `spec/controllers/omniauth_callbacks_controller_spec.rb`.
- Tests have been added to account for the many potential user scenarios when signing in with or linking external credentials.
- In addition, the following previous problems/silent errors have been addressed:
  - Remove redefining of `User.from_omniauth` in tests
  - Fix user creation in omniauth tests
    - Prior to this change, `let!(:user) { User.create!(email: 'user@organization.ca', firstname: 'John', surname: 'Doe', org: @org) }` (bang (!) added)  would fail with the following error:

	```
	Failure/Error: let!(:user) { User.create!(email: 'user@organization.ca', firstname: 'John', surname: 'Doe', org: @org) }

     	  ActiveRecord::RecordInvalid:
       	    Validation failed: Password can't be blank
	```

momo3404 commented on spec/controllers/omniauth_callbacks_controller_spec.rb:
```diff
@@ -2,131 +2,193 @@
+                           '<a href="mailto:dmp-assistant@tech.alliancecan.ca">dmp-assistant@tech.alliancecan.ca</a>'
+  }
+}.freeze
+FLASH_MESSAGES[:alert][:missing_email_link] = FLASH_MESSAGES[:alert][:missing_email_sign_in].sub('sign in', 'link')
```
Why did you opt to add `missing_email_link` to `FLASH_MESSAGES` this way instead of also assigning it under `alert`?

aaronskiba replied:
I guess I just wanted the fewer lines of code. I guess it makes sense too, because the two should be identical other than the `sign in` vs `link` difference.

momo3404 commented on spec/controllers/omniauth_callbacks_controller_spec.rb:
```diff
@@ -2,131 +2,193 @@
     create(:org, managed: false, is_other: true)
-    @org = create(:org, managed: true)
     @identifier_scheme = create(:identifier_scheme, :openid_connect)
     @request.env['omniauth.auth'] = OmniAuth.config.mock_auth[:openid_connect].dup
```
just curious- why use `.dup` is used here? Is it so that each test has its own mock authentication?

aaronskiba replied:
Yes, I want a fresh copy of `OmniAuth.config.mock_auth[:openid_connect]` across each test. Without `.dup`, any modifications to  ` @request.env['omniauth.auth']` (e.g. `@request.env['omniauth.auth'].info.email = nil`) would persist across tests.

momo3404 commented on spec/controllers/omniauth_callbacks_controller_spec.rb:
```diff
@@ -2,131 +2,193 @@
+            expect(user.email).to eq(@request.env['omniauth.auth'].info.email)
+            expect(user.firstname).to eq(@request.env['omniauth.auth'].info.first_name)
+            expect(user.surname).to eq(@request.env['omniauth.auth'].info.last_name)
+            expect(user.org).to eq(Org.find_by(is_other: true))
```
just checking, is `Org.find_by(is_other: true)` returning a fallback organization?

aaronskiba replied:
`create(:org, managed: false, is_other: true)` is being executed before this, but I could still add the bang operator.

aaronskiba replied:
Or maybe I'm not fully understanding your question. I believe there is currently only one org such that org.is_other == true. However, maybe we should eventually update the code to handle the is_other org in the same manner we handle the app's default org.

momo3404 replied:
Mb, I didn't remember the `create` with `is_other` set to true. I get what it's testing now

momo3404 commented on spec/controllers/omniauth_callbacks_controller_spec.rb:
```diff
@@ -2,131 +2,193 @@
+            user.reload
+          end.to change(user.identifiers, :count).by(1)
+          expect(User.count).to eq(1)
+          expect(Identifier.count).to eq(1)
```
Do you think the check of `identifier.count` is necessary if it's already checked in line 102?

aaronskiba replied:
Maybe it's a little bit much, but still not totally redundant. I guess the fact that user.identifiers.count changed by one, does not necessarily mean that there is only one Identifier object. It might be okay to leave as is.

momo3404 replied:
Yea I agree, not redundant enough to warrant changing

momo3404 commented on spec/controllers/omniauth_callbacks_controller_spec.rb:
```diff
@@ -2,131 +2,193 @@
+        end
       end
     end
+
```
Do these cases cover the case of repeated sign-in/linking attempts with the same credentials? That could create duplicate identifiers. If they do and I missed it then feel free to quickly resolve this

aaronskiba replied:
This could be good to add. I'll look into it.

aaronskiba replied:
There is a custom validator for this in the Identifier model. I'll add a test to ensure it's getting triggered.

`app/models/identifier.rb`

```ruby
  # Ensure that the identifiable only has one identifier for the scheme
  def value_uniqueness_with_scheme
    if new_record? && Identifier.where(identifier_scheme: identifier_scheme,
                                       identifiable: identifiable).any?
      errors.add(:identifier_scheme, _('already assigned a value'))
    end
  end
```

aaronskiba replied:
Or maybe this is already covered in `spec/models/identifier_spec.rb`:

```ruby
      it 'prevents multiple identifiers per identifier_scheme' do
        scheme = create(:identifier_scheme)
        create(:identifier, identifiable: @org, identifier_scheme: scheme,
                            value: Faker::Lorem.word)
        id = build(:identifier, identifiable: @org, identifier_scheme: scheme,
                                value: Faker::Number.number.to_s)
        byebug
        expect(id.valid?).to eql(false)
        expect(id.errors[:identifier_scheme].present?).to eql(true)
      end
```
```
(byebug) id.valid?
false
(byebug) id.errors[:identifier_scheme]
["already assigned a value"]
```

aaronskiba replied:
Testing the same thing within `spec/controllers/omniauth_callbacks_controller_spec.rb` (added `byebug` and `!` to `create!()`):

```ruby
    def create_user_identifier_from_auth(user)
      byebug
      Identifier.create(identifier_scheme: @identifier_scheme,
                        value: request.env['omniauth.auth'].uid,
                        attrs: request.env['omniauth.auth'],
                        identifiable: user)
    end
```
```
(byebug) Identifier.create!(identifier_scheme: @identifier_scheme, value: request.env['omniauth.auth'].uid, attrs: request.env['omniauth.auth'], identifiable: user)
#<Identifier id: 1, value: "https://www.cilogon.org/12345", attrs: "{\"provider\":\"openid_connect\",\"uid\":\"https://www.ci...", identifier_scheme_id: 1, identifiable_id: 1, identifiable_type: "User", created_at: "2025-06-19 18:47:36.229071974 +0000", updated_at: "2025-06-19 18:47:36.229071974 +0000">
(byebug) Identifier.create!(identifier_scheme: @identifier_scheme, value: request.env['omniauth.auth'].uid, attrs: request.env['omniauth.auth'], identifiable: user)
*** ActiveRecord::RecordInvalid Exception: Validation failed: Identifier scheme already assigned a value

nil
```

aaronskiba committed:
Add `!` to OmniauthCallbacksController tests

aaronskiba committed:
Handle edge case in OmniauthCallbacksController

Added an `else` condition within OmniauthCallbacksController#handle_openid_connect_for_signed_in_user to handle when `user.id == current_user.id`.
- As explained in the added comments, "This scenario is not possible through normal UI interaction, but a user could trigger it by manipulating the frontend."
- Added a test for this within `spec/controllers/omniauth_callbacks_controller_spec.rb` as well.

aaronskiba committed:
Add tests & update `translations_for_all_locales`

`spec/services/locale_service_spec.rb`
- Set `I18n.available_locales = %i[en-CA fr-CA]` within tests, to specifically match DMP Assistant
- Added tests for some edge case

`app/services/locale_service.rb`
- Updated `LocaleService.translations_for_all_locales` to handle the edge cases tests within the corresponding `_spec.rb` file.

aaronskiba committed:
Update CHANGELOG.md

momo3404 reviewed (approved):
Great work with fixing this and improving the tests!
